File: scan-shortcut.js
/*
scan-shortcut.js

This exports a function that accepts an array of actions obtained by shortcuts.js.
It returns an alphabetized list of all the unique permissions in the shortcut.
Currently it ignores third-party apps' actions, though that may change in the future.
*/

const axios = require('axios').default;
const bplist = require('./node_modules/bplist-parser/');

const actionList = require('./actions.json');

module.exports = async (shortcut) => {
    const actions = bplist.parseBuffer(shortcut)[0].WFWorkflowActions;

    const perms = [];
    
    for (const action of actions) {
        actionID = action.WFWorkflowActionIdentifier
        if (actionList.hasOwnProperty(actionID)) {
            for (const perm of actionList[actionID]) {
                if (!perms.includes(perm)) perms.push(perm);
            }
        }
    }

    return perms.sort();
}

File: scan_shortcut.py
import json
import plistlib

action_list = json.loads(open("./actions.json").read())

def scan_shortcut(plist):
    '''Accepts a bytes object containing a shortcut, and returns a list of the shortcut's required permissions.'''
    
    actions = plistlib.loads(plist)["WFWorkflowActions"]

    perms = []

    for action in actions:
        action_id = action["WFWorkflowActionIdentifier"]
        if action_id in action_list:
            for perm in action_list[action_id]:
                if not perm in perms:
                    perms.append(perm)
    
    perms.sort()
    return perms